nixos/modules/security/doas.nix

```diff
@@ -0,0 +1,259 @@
+                list.
+              '';
+            };
+
+            users = mkOption {
+              type = with types; listOf (either str int);
+              default = [];
+              description = "The usernames / UIDs this rule should apply for.";
+            };
+
+            groups = mkOption {
+              type = with types; listOf (either str int);
+              default = [];
+              description = "The groups / GIDs this rule should apply for.";
+            };
+
+            runAs = mkOption {
+              type = with types; nullOr str;
+              default = null;
+              description = ''
+                Under which user/group the specified command is allowed to run.
+
+                A user can be specified using just the username: <code>"foo"</code>. It is also
+                possible to only allow running as a specific group with <code>":bar"</code>.
+              '';
+            };
+
+            cmd = mkOption {
+              type = with types; nullOr str;
+              default = null;
+              description = ''
+                The command the user is allowed to run. The default is all commands. Be advised that
+                it is best to specify absolute paths. If a relative path is specified, only a
+                restricted PATH will be searched. When set to <code>null</code>, all commands are
+                allowed.
+              '';
+            };
+
+            args = mkOption {
+              type = with types; nullOr (listOf str);
+              default = [];
+              description = ''
+                Arguments to the command. The arguments provided by the user need to match those
+                specified. When set to <code>null</code>, the command must be run without any
+                arguments.
+              '';
+            };
+          };
+        }
+      );
+    };
+
+    extraConfig = mkOption {
+      type = with types; lines;
+      default = "";
+      description = ''
+        Extra configuration text appended to <filename>doas.conf</filename>.
+      '';
+    };
+  };
+
+
+  ###### implementation
+
+  config = mkIf cfg.enable {
+
+    security.doas.extraRules = [
+      {
+        groups = [ "wheel" ];
+        noPass = !cfg.wheelNeedsPassword;
+      }
+    ];
+
+    security.wrappers = {
+      doas.source = "${doas}/bin/doas";
+    };
+
+    environment.systemPackages = [
+      doas
+    ];
+
+    security.pam.services.doas = {
+      allowNullPassword = true;
+      sshAgentAuth = true;
+    };
+
+    environment.etc."doas.conf" = {
+      source = pkgs.runCommand "doas-conf"
+        {
+          src = pkgs.writeText "doas-conf-in" ''
+            # Do NOT edit this file by hand. Set the NixOS options
+            # ‘security.doas.extraRules’ or ‘security.doas.extraConfig’ instead.
+            # In order to completely replace the contents of this file, use
+            # ‘environment.etc."doas.conf"’.
```

Comment:
Surely this can’t be edited by hand? The store is mounted read-only on NixOS.

Reply:
Not even as root?

The main reason this is here is because `sudo` has a similar warning in its module. This is easily rectified, though, if you think it's unnecessary.

Reply:
Considering the option `nix.readOnlyStore` exists (whether or not it should is another conversation), I think I'll keep this here. If there's even a *chance* someone tries to edit it directly, I'd rather there be a warning so I can't be held liable.

Nonetheless, I appreciate your feedback; thank you!

Reply:
Even as root. If somebody has set readOnlyStore to false, I think they’ve purposefully decided to shoot themselves in the foot. :) We can’t copy this warning to every file In /etc. I think maybe the reason sudo has it is because there’s usually a special flow to editing the sudoers file (visudo). I don’t know if doas does similar.

Anyway, if you really want to keep the comment, it’s not going to get in anybody’s way.

Reply:
Just as an FYI, I exchanged the "do not edit this file by hand" for "To modify this file...".